ppcs/ppc0023-lexical-require.md

```diff
@@ -0,0 +1,152 @@
+# Lexically require modules
+
+## Preamble
+
+    Author:  Ovid
+    Sponsor:
+    ID:      23 
+    Status:  Draft
+
+## Abstract
+
+When writing a module, the `use` and `require` statements make the modules
+available globally. This leads to strange bugs where you can write
+`my $object = Some::Class->new` and have it work, even if you didn't
+require that module. This transitive dependency is fragile and will break
+if the code requiring `Some::Class` decides to no longer require it.
+
+This PPC proposes:
+
+```perl
+package Foo;
+use feature 'lexical_require`;
+use Some::Module;
+```
+
+With the above, code outside of the above scope cannot see `Some::Module` unless
```

Comment:
I think @Grinnz is quite fairly calling out this as confusing or ambiguous.  It should probably be worded:

```
Within the lexcial scope of the 'lexical_require' feature code cannot call methods against class names that have not been explicitly required within the current package, and doing so would throw an exception.  Methods would be allowed against any object (blessed reference), but not against a class.
```

Reply:
@demerphq suggested:

> Within the lexical scope of the 'lexical_require' feature, code cannot call methods against class names that have not been explicitly required within the current package, and doing so would throw an exception. Methods would be allowed against any object (blessed reference), but not against a class.

I read that as changing the meaning, but I'm unsure if it was meant to. Rewriting it:

> Within a given lexical scope, **ß**, if the 'lexical_require' feature is used, code outside of scope **ß** cannot call methods against class names that have not been explicitly required within the current package, and doing so would throw an exception. Methods would be allowed against any object (blessed reference), but not against a class.

However, that doesn't mean the transitive dependencies aren't available. If scope **ß** uses `lexical_require` and `Hash::Ordered`, but scope **∂** uses `Hash::Ordered` but _doesn't_ use `lexical_require`, then `Hash::Ordered` is still available to everyone as a transitive dependency.

Thus, this doesn't offer perfect protection, but it does mean that I can write my own code in such a way that people will be less likely to rely on my internals.

Reply:
OK, now I know I'm understanding you correctly. I'll rewrite the PPC as soon as I practically can. Makes total sense.